docs/proposals/runtime-client-server.md

```diff
@@ -0,0 +1,221 @@
+    rpc ListImages(ListImagesRequest) returns (ListImagesResponse) {}
+    // ImageStatus returns the status of the image.
+    rpc ImageStatus(ImageStatusRequest) returns (ImageStatusResponse) {}
+    // PullImage pulls a image with authentication config.
+    rpc PullImage(PullImageRequest) returns (PullImageResponse) {}
+    // RemoveImage removes the image.
+    rpc RemoveImage(RemoveImageRequest) returns (RemoveImageResponse) {}
+}
+```
+
+Note that some types in [pkg/kubelet/api/v1alpha1/runtime/api.proto](../../pkg/kubelet/api/v1alpha1/runtime/api.proto) are already defined at [Container runtime interface/integration](https://github.com/kubernetes/kubernetes/pull/25899).
+We should decide how to integrate the types in [#25899](https://github.com/kubernetes/kubernetes/pull/25899) with gRPC services:
+
+* Auto-generate those types into protobuf by [go2idl](../../cmd/libs/go2idl/)
+  - Pros:
+    - trace type changes automatically, all type changes in Go will be automatically generated into proto files
+  - Cons:
+    - type change may break existing API implementations, e.g. new fields added automatically may not noticed by remote runtime
+    - needs to convert Go types to gRPC generated types, and vise versa
+    - needs processing attributes order carefully so as not to break generated protobufs (this could be done by using [protobuf tag](https://developers.google.com/protocol-buffers/docs/gotutorial))
+    - go2idl doesn't support gRPC, [protoc-gen-gogo](https://github.com/gogo/protobuf) is still required for generating gRPC client
+* Embed those types as raw protobuf definitions and generate Go files by [protoc-gen-gogo](https://github.com/gogo/protobuf)
+  - Pros:
+    - decouple type definitions, all type changes in Go will be added to proto manually, so it's easier to track gRPC API version changes
+    - Kubelet could reuse Go types generated by `protoc-gen-gogo` to avoid type conversions
+  - Cons:
+    - duplicate definition of same types
+    - hard to track type changes automatically
+    - need to manage proto files manually
+
+For better version controlling and fast iterations, this proposal embeds all those types in `api.proto` directly.
+
+## Implementation
+
+Each new runtime should implement the [gPRC](http://www.grpc.io) server based on [pkg/kubelet/api/v1alpha1/runtime/api.proto](../../pkg/kubelet/api/v1alpha1/runtime/api.proto). For version controlling, `KubeletGenericRuntimeManager` will request `RemoteRuntime`'s `Version()` interface with the runtime api version. To keep backward compatibility, the API follows standard [protobuf guide](https://developers.google.com/protocol-buffers/docs/proto) to deprecate or add new interfaces.
+
+A new flag `--container-runtime-endpoint` (overrides `--container-runtime`) will be introduced to kubelet which identifies the endpoint of the remote runtime. And new flag `--image-service-endpoint` will be introduced to kubelet which identifies the endpoint of the image service.
```

Comment:
Didn't consider this way before. One difference from Docker drivers is that container runtime in kubelet is usually not changed after deployment and only one runtime is supported now, while Docker drivers often support multiple one.

Added it to #28642 for more discussion since this page is hard to open for most.